pyvisa_py/tcpip.py

```diff
@@ -41,6 +41,203 @@
 
 @Session.register(constants.InterfaceType.tcpip, "INSTR")
 class TCPIPInstrSession(Session):
+    """A class to dispatch to VXI11 or HiSLIP based on the protocol."""
+
+    def __new__(
+        cls,
+        resource_manager_session: VISARMSession,
+        resource_name: str,
+        parsed=None,
+        open_timeout: Optional[int] = None,
+    ):
+        obj: Union[TCPIPInstrHiSLIP, TCPIPInstrVxi11]
+
+        if parsed is None:
+            parsed = rname.parse_resource_name(resource_name)
+
+        if parsed.lan_device_name.lower().startswith("hislip"):
+            obj = TCPIPInstrHiSLIP(
+                resource_manager_session, resource_name, parsed, open_timeout
+            )
+        else:
+            obj = TCPIPInstrVxi11(
+                resource_manager_session, resource_name, parsed, open_timeout
+            )
+        return obj
```

Comment:
We could allow TCPIPInstrHiSLIP, TCPIPInstrVxi11 to subclass this class by modifying the method as such:
```suggestion
    def __new__(
        cls,
        resource_manager_session: VISARMSession,
        resource_name: str,
        parsed=None,
        open_timeout: Optional[int] = None,
    ):
        if cls is not TCPIPInstrSession:
            return super().__new__(cls, resource_manager_session, resource_name, parsed, open_timeout)

        if parsed is None:
            parsed = rname.parse_resource_name(resource_name)

        if parsed.lan_device_name.lower().startswith("hislip"):
            cls = TCPIPInstrHiSLIP
        else:
            cls = TCPIPInstrVxi11
            
        return cls(resource_manager_session, resource_name, parsed, open_timeout)
```

This would allow to share common parts more easily.

Reply:
If we make TCPIPInstrHiSLIP and TCPIPInstrVxi11 subclass TCPIPInstrSession instead of Session, we add a couple of complications for absolutely no benefit.

The first complication is having to test if cls is TCPIPInstrSession, then calling super().__new__().  (And that didn't work for me, I had to call Session.__new__() instead.)

The second complication is that the return value has to be cls.__new__() instead of cls().  Otherwise __init__ gets called twice on the resulting object. 

After trying it both ways, my preference is for TCPIPInstrHiSLIP and TCPIPInstrVxi11 to subclass Session.  It just feels like there's less dunder magic going on.  But it will work either way.

Reply:
It looks like everywhere I typed \_\_new\_\_ it got replaced with a bold new.

Reply:
My main concern was to avoid code duplication between VXI11 and HiSLIP since the interface object in both cases can have similar APIs but I may be over complicating things here. So let's just keep your current solution.